### ISSUE TITLE
Inline WHERE clauses appear to create a concrete class of entities in order to compile

### ISSUE THREAD

**Issue by benliebert** [linq2db/linq2db#1099]
Here's something curious.  I have abstracted my LINQ queries so that they run against interfaces instead of concrete classes.  Deep down in my code when it queries for (in this case `cxt.BackgroundTask` - see image below), I return a list of concrete class `BackgroundTask : IBackgroundTask`.  This system works beautifully - for reads and writes.

However, I just found that it fails when you have an embedded WHERE clauses, as below:

![image](https://user-images.githubusercontent.com/2273767/38704184-9f225242-3ef9-11e8-8b33-772d3b23456e.png)

You can see from the comments that if I include the WHERE clause within the initial query, it fails because it tries to create an _instance_ of the `IBackgroundTask`:

> Type 'IBackgroundTask' does not have a default constructor

However, if I move the WHERE clause to an outer scope and use a lambda, it works fine.  

It wouldn't be too much of a problem for me to update all my queries to use lambdas instead, however it's a tough thing for my team to remember and I'm curious as to what is going on?

```
Exception message:
Stack trace:
"   at System.Linq.Expressions.Expression.New(Type type)\r\n   at LinqToDB.Linq.Builder.TableBuilder.TableContext.BuildDefaultConstructor(EntityDescriptor entityDescriptor, Type objectType, Int32[] index) in c:\\projects\\linq2db\\Source\\Linq\\Builder\\TableBuilder.cs:line 324\r\n   at LinqToDB.Linq.Builder.TableBuilder.TableContext.BuildTableExpression(Boolean buildBlock, Type objectType, Int32[] index) in c:\\projects\\linq2db\\Source\\Linq\\Builder\\TableBuilder.cs:line 298\r\n   at LinqToDB.Linq.Builder.TableBuilder.TableContext.BuildQuery[T](Query`1 query, ParameterExpression queryParameter) in c:\\projects\\linq2db\\Source\\Linq\\Builder\\TableBuilder.cs:line 615\r\n   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]() in c:\\projects\\linq2db\\Source\\Linq\\Builder\\ExpressionBuilder.cs:line 169\r\n   at LinqToDB.Linq.Query`1.CreateQuery(IDataContext dataContext, Expression expr) in c:\\projects\\linq2db\\Source\\Linq\\Query.cs:line 261\r\n   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr) in c:\\projects\\linq2db\\Source\\Linq\\Query.cs:line 210\r\n   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache) in c:\\projects\\linq2db\\Source\\Linq\\ExpressionQuery.cs:line 89\r\n   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator() in c:\\projects\\linq2db\\Source\\Linq\\ExpressionQuery.cs:line 186\r\n   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)\r\n   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)\r\n   at Foundation.BackgroundTasks.BackgroundTaskManager.GetTasks(Nullable`1 personId, Boolean includeCompleted, Boolean includeDeleted) in ...
```

### Environment details
linq2db version: *1.9.0*  
Database Server:  Sql Server 2017  
Operating system: Windows 10
Framework version: .NET Framework 4.6.1

## More code/info

    public partial class BackgroundTask : IBackgroundTask
	    {
		    public int CalendarTypeID { get; set; } = (int)CalendarTypes.FoundationBackgroundTask;
		    public int? ID { get { return this.BackgroundTaskID; } }
    
		    public int? ClientID {
		    	get { return this.AccountID; }
		    	set { this.AccountID = value; }
		    }
    
		    public bool IsSystemTask
		    {
		    	get { return !this.PersonID.HasValue; }
		    }
    
		    public int DurationID { get; set; }
		    public int DurationInterval { get; set; }
    
	    }

    public class CoreRepositoriesInitializer : IBackgroundTaskRepositoryInitializer{  
    		public IBackgroundTaskRepository CreateBackgroundTaskRepository()
    		{
    			var rep = Dependency.Resolve<IRepositoryInitializer<ClientRepository>>();
    			var result = rep.Create();
    			return result;
    		}
    }

    public partial interface ClientRepository : IRepository
    {
    		IQueryable<APIContracts.Entities.Data.BackgroundTask> BackgroundTask { get; }
    }

    public partial class Linq2Db_Client : DataConnection, IRepository, ClientRepository
    	{
    		public virtual IQueryable<APIContracts.Entities.Data.BackgroundTask> BackgroundTask {
    			get { 
    				return this.GetTable<BackgroundTask>();
    			}
    		}
    
	    	IQueryable<IBackgroundTask> IBackgroundTaskRepository.BackgroundTask
	    	{
	    		get { return BackgroundTask; }
	    	}
    }

**Comment by sdanyliv** [comment]
Which type has ctx.BackgroundTask?
Can you reproduce it in small test case?

**Comment by benliebert** [comment]
@sdanyliv I've added some extra code in my original post.  It seems rather elaborate because we've abstracted the _queries_ away from the underlying data types (hence the use of interfaces), but it's all coming together beautifully - just this little weird niggle

**Comment by sdanyliv** [comment]
Fixed in #1106
